### PR TITLE
BGP-45 [Frontend] Logout, router protector, register input validation

### PR DESCRIPTION
**Pull request by ZichanYang324**
# Description
Modify some detailed features of the login page and register page. Fix the bugs in the last PR.

## Related PRS (if any):

This is following on the [PR#19](https://github.com/ZichanYang324/BF-Employee-HR/pull/19) and [PR#23](https://github.com/ZichanYang324/BF-Employee-HR/pull/23)

## Main changes explained:
- Add input validation in the register page
- Add router protector
- Add the logout feature

## How to test:
1. check into the current branch
2. do `npm install` and `npm run dev` to run this PR locally(on both client side and server side)
3. navigate to `.../register` to test the input validation
4. when finished registering, it will automatically log in and then click on the top right corner to test the logout
5. test the router protector, before login, tried to navigate to the `.../dashboard`
6. test the login page

## Screenshots or videos of changes or expected result:
Screenshot 2024-02-26 at 21 50 46
Screenshot 2024-02-26 at 21 49 26
Screenshot 2024-02-26 at 22 16 40

## Note:
N/A

### CODE REVIEW

File: Employee-Client/src/components/login/index.jsx [hunk 1]
```diff
@@ -13,29 +12,49 @@ import Typography from '@mui/material/Typography';
...
+        if (usernameInput.isValid && passwordInput.isValid) {
+            setSubmitIsValid(true)
+        } else {
+            setSubmitIsValid(false)
+        }
+    }
+
+    useEffect(() => {
+        checkPgae()
+    }, [usernameInput.value, passwordInput.value])
```
Comment (yuelikestocode) [inline Employee-Client/src/components/login/index.jsx:56]: Is it better to have 
```suggestion
    const checkPgae = () => {
        if (usernameInput.isValid && passwordInput.isValid) {
            setSubmitIsValid(true)
        } else {
            setSubmitIsValid(false)
        }
    }

    useEffect(() => {
        checkPgae()
    }, [usernameInput.isValid, passwordInput.isValid])
```

where `[usernameInput.value, passwordInput.value]` are not directly used in `checkPgae`

File: Employee-Client/src/components/login/index.jsx [hunk 2]
```diff
@@ -13,29 +12,49 @@ import Typography from '@mui/material/Typography';
...
         event.preventDefault();
-        // customFetch.post('/user/auth', { username: username, password: password })
-        // .then((res) => {
-        //     console.log(res);
-        // })
-        dispatch(loginUser({ username: username, password: password }));
+        dispatch(loginUser({ username: usernameInput.value, password: passwordInput.value }));
     };
 
+    const checkPgae = () => {
```
Comment (yuelikestocode) [inline Employee-Client/src/components/login/index.jsx:46]: Nit pick: typo

**Comment by ZichanYang324** [comment]
@yuelikestocode I have took in your advice and change the login/index, please check, sorry about the typo problem.

**Review by christorange — APPROVED** [review]
LGTM, functionalities all work well. Since we are storing auth token in local storage we don't necessarily need a sign out API, but it would be more	secure if we have a blacklist on db, add deprecated tokens to them, and make sure the token is not in blacklist everytime a user signs in.

**Comment by ZichanYang324** [comment]
@christorange Since this is not list on the project requirements, then I think I will not do it this time, sorry about that.
I need to ensure all the requirements features can be finished on time. And if we can finished ahead of the ddl and still have some time, then I will tried to get it done.

**Pull request merged**